File: Leetcode/152. Maximum Product Subarray/MaximumProductSubarray.cpp
// Leetcode 152. Maximum Product Subarray
// https://leetcode.com/problems/maximum-product-subarray/description/
// Runtime: 152ms (...)

// brutal-force, need optimization
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }

        int res = INT_MIN;
        
        int *maxarr = new int[nums.size()];
        for (int i = 0; i < nums.size(); i++) {
            maxarr[i] = INT_MIN;
        }
        
        for (int i = 0; i < nums.size(); i++) {
            int max = nums[i];
            if (max > maxarr[i]) {
                maxarr[i] = max;
            }
            
            for (int j = i + 1; j < nums.size(); j++) {
                max = max * nums[j];
                if (max > maxarr[j]) {
                    maxarr[j] = max;
                }
            }
        }
        
        for (int i = 0; i < nums.size(); i++) {
            if (maxarr[i] > res) {
                res = maxarr[i];
            }
        }
        
        return res;
    }
};

File: Leetcode/152. Maximum Product Subarray/maximum_product_subarray.rs
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut result: i32 = i32::MIN;
        let mut product: i32 = 1;
        for num in nums.iter() {
            if product == 0 {
                product = 1;
            }
            product *= *num;
            result = std::cmp::max(result, product);
        }
        
        product = 1;
        for num in nums.iter().rev() {
            if product == 0 {
                product = 1;
            }
            
            product *= *num;
            result = std::cmp::max(result, product);
        }
        
        return result;
    }
}
